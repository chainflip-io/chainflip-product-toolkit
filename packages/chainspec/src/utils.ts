import assert from 'assert';
import * as prettier from 'prettier';
import { type ParsedMetadata } from './BaseParser';

type ChangeType = 'added' | 'removed' | 'changed';
type Change = { path: string[]; type: ChangeType };

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(a: any, b: any, path: string[] = []): Generator<Change> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

class ChangelogGenerator {
  addedOrRemovedPallets: Record<string, 'added' | 'removed'> = {};

  changedEvents: Record<
    `${string}.${string}`,
    { field: string; type: 'added' | 'changed' | 'removed' }[]
  > = {};
  addedOrRemovedEvents: Record<`${string}.${string}`, 'added' | 'removed'> = {};

  trackedFieldChanges = new Set<string>();

  add({ path, type }: Change) {
    if (path.length === 1) {
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedPallets[path[0]] = type;
    } else if (path.length === 2) {
      const name = `${path[0]}.${path[1]}` as const;
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedEvents[name] = type;
    } else {
      const [pallet, event, , field] = path;
      const id = `${pallet}.${event}.${field}`;
      if (this.trackedFieldChanges.has(id)) return;
      this.trackedFieldChanges.add(id);
      (this.changedEvents[`${pallet}.${event}`] ??= []).push({ field, type });
    }
  }

  toString() {
    const lines: string[] = [];

    const pallets = Object.keys(this.addedOrRemovedPallets).sort();

    if (pallets.length) {
      lines.push('New or removed pallets:');
      for (const pallet of pallets) {
        lines.push(`  ${pallet}: ${this.addedOrRemovedPallets[pallet]}`);
      }
      lines.push('');
    }

    const events = Object.keys(this.addedOrRemovedEvents)
      .concat(Object.keys(this.changedEvents))
      .sort() as `${string}.${string}`[];

    let currentPallet = '';
    for (const event of events) {
      const [pallet, name] = event.split('.');
      if (pallet !== currentPallet) {
        if (currentPallet) lines.push('');
        currentPallet = pallet;
        lines.push(`${pallet}:`);
      }
      if (event in this.addedOrRemovedEvents) {
        lines.push(`  - ${name}: ${this.addedOrRemovedEvents[event]}`);
      } else {
        lines.push(`  - ${name}:`);
        const fields = this.changedEvents[event].map(
          ({ field, type }) => `    - ${field}: ${type}`,
        );
        lines.push(...fields);
      }
    }

    lines.push('');
    return lines.join('\n');
  }
}

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();
  const changelog = new ChangelogGenerator();

  for (const change of diff(a, b)) {
    changelog.add(change);

    const [pallet, event, field] = change.path;

    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return { changedOrAddedEvents, changelog: changelog.toString() };
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });
