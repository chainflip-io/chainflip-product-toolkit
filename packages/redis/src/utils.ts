type SnakeToCamelCase<T extends string> = T extends `${infer P}_${infer S}`
  ? `${P}${Capitalize<SnakeToCamelCase<S>>}`
  : T;

export const snakeToCamelCase = <const T extends string>(str: T): SnakeToCamelCase<T> =>
  str.replace(/_(.)/g, (match, group: string) => group.toUpperCase()) as SnakeToCamelCase<T>;

type CamelCasedKeys<T> =
  T extends Record<string, unknown>
    ? { [K in keyof T as K extends string ? SnakeToCamelCase<K> : K]: CamelCasedKeys<T[K]> }
    : T extends (infer U)[]
      ? CamelCasedKeys<U>[]
      : T;

export const transformKeysToCamelCase = <T>(obj: T): CamelCasedKeys<T> => {
  if (obj == null || typeof obj !== 'object') return obj as CamelCasedKeys<T>;

  if (Array.isArray(obj)) return obj.map(transformKeysToCamelCase) as CamelCasedKeys<T>;

  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => [
      snakeToCamelCase(key),
      transformKeysToCamelCase(value),
    ]),
  ) as CamelCasedKeys<T>;
};
