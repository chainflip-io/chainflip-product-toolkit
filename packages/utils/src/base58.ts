import { decodeBytesWithCharset, encodeBytesWithCharset } from './bytes';

export const CHARSET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

export const encode = (bytes: Uint8Array | number[]): string =>
  encodeBytesWithCharset(bytes, CHARSET);

export const decode = (input: string): Uint8Array => decodeBytesWithCharset(input, CHARSET);
