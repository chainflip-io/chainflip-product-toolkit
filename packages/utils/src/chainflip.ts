export const chainflipAssets = [
  'Usdc',
  'Usdt',
  'Flip',
  'Dot',
  'Eth',
  'Btc',
  'ArbUsdc',
  'ArbEth',
  'Sol',
  'SolUsdc',
] as const;

export type ChainflipAsset = (typeof chainflipAssets)[number];

export const rpcAssets = ['USDC', 'USDT', 'FLIP', 'DOT', 'ETH', 'BTC', 'SOL'] as const;
type RpcAsset = (typeof rpcAssets)[number];

export type BaseChainflipAsset = Exclude<ChainflipAsset, 'Usdc'>;

export const baseChainflipAssets = chainflipAssets.filter(
  (asset): asset is BaseChainflipAsset => asset !== 'Usdc',
);

export const chainflipEvmChains = ['Ethereum', 'Arbitrum'] as const;
export type ChainflipEvmChain = (typeof chainflipEvmChains)[number];

export const chainflipChains = [...chainflipEvmChains, 'Bitcoin', 'Polkadot', 'Solana'] as const;
export type ChainflipChain = (typeof chainflipChains)[number];

export const chainflipNetworks = ['backspin', 'sisyphos', 'perseverance', 'mainnet'] as const;
export type ChainflipNetwork = (typeof chainflipNetworks)[number];

export type AssetOfChain<C extends ChainflipChain> = (typeof chainConstants)[C]['assets'][number];

export type ChainAssetMap<T> = {
  [C in ChainflipChain]: {
    [A in (typeof assetConstants)[AssetOfChain<C>]['rpcAsset']]: T;
  };
};
export const readAssetValue = <T>(value: ChainAssetMap<T>, asset: ChainflipAsset): T => {
  const chainValues = value[assetConstants[asset].chain];
  return chainValues[assetConstants[asset].rpcAsset as keyof typeof chainValues];
};

export const assetConstants = {
  Eth: {
    chain: 'Ethereum',
    rpcAsset: 'ETH',
    decimals: 18,
  },
  Flip: {
    chain: 'Ethereum',
    rpcAsset: 'FLIP',
    decimals: 18,
  },
  Usdc: {
    chain: 'Ethereum',
    rpcAsset: 'USDC',
    decimals: 6,
  },
  Usdt: {
    chain: 'Ethereum',
    rpcAsset: 'USDT',
    decimals: 6,
  },
  Dot: {
    chain: 'Polkadot',
    rpcAsset: 'DOT',
    decimals: 10,
  },
  Btc: {
    chain: 'Bitcoin',
    rpcAsset: 'BTC',
    decimals: 8,
  },
  ArbUsdc: {
    chain: 'Arbitrum',
    rpcAsset: 'USDC',
    decimals: 6,
  },
  ArbEth: {
    chain: 'Arbitrum',
    rpcAsset: 'ETH',
    decimals: 18,
  },
  Sol: {
    chain: 'Solana',
    rpcAsset: 'SOL',
    decimals: 9,
  },
  SolUsdc: {
    chain: 'Solana',
    rpcAsset: 'USDC',
    decimals: 6,
  },
} as const satisfies Record<
  ChainflipAsset,
  {
    chain: ChainflipChain;
    rpcAsset: RpcAsset;
    decimals: number;
  }
>;

export const chainConstants = {
  Ethereum: {
    assets: ['Eth', 'Flip', 'Usdc', 'Usdt'],
    gasAsset: 'Eth',
    blockTimeSeconds: 12,
  },
  Polkadot: {
    assets: ['Dot'],
    gasAsset: 'Dot',
    blockTimeSeconds: 6,
  },
  Bitcoin: {
    assets: ['Btc'],
    gasAsset: 'Btc',
    blockTimeSeconds: 10 * 60,
  },
  Arbitrum: {
    assets: ['ArbUsdc', 'ArbEth'],
    gasAsset: 'ArbEth',
    blockTimeSeconds: 0.26,
  },
  Solana: {
    assets: ['Sol', 'SolUsdc'],
    gasAsset: 'Sol',
    blockTimeSeconds: 0.8,
  },
} as const satisfies Record<
  ChainflipChain,
  {
    assets: ChainflipAsset[];
    gasAsset: ChainflipAsset;
    blockTimeSeconds: number;
  }
>;
