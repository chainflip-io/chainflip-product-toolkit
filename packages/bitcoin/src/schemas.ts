import { hexToBytes } from '@chainflip/utils/bytes';
import BigNumber from 'bignumber.js';
import { z } from 'zod';

const hexString = z.string().regex(/^([0-9a-f]{2})+$/, { message: 'expected hex string' });

const vout = z.object({
  value: z.number().transform((n) => BigInt(new BigNumber(n).shiftedBy(8).toFixed(0))),
  n: z.number(),
});

const nulldataVout = z
  .object({
    scriptPubKey: z.object({
      type: z.literal('nulldata'),
      // remove OP_RETURN and PUSH_BYTES_XX
      hex: hexString.transform((x) => hexToBytes(`0x${x.slice(4)}`)),
    }),
  })
  .and(vout);

const addressVout = z
  .object({
    scriptPubKey: z.object({
      type: z.enum([
        'witness_v1_taproot',
        'witness_v0_scripthash',
        'witness_v0_keyhash',
        'pubkeyhash',
        'scripthash',
      ]),
      address: z.string(),
    }),
  })
  .and(vout);

export const txSchema = z.object({
  vout: z.tuple([addressVout, nulldataVout, addressVout]),
  blockhash: hexString,
});

export const blockSchema = z.object({
  height: z.number(),
});
