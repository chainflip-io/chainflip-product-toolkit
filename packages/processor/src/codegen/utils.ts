import { Queue } from '@chainflip/utils/async';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as prettier from 'prettier';
import { z } from 'zod';
import { ParsedMetadata } from './Parser';
import assert from 'assert';

type ChangeType = 'added' | 'removed' | 'changed';
type Change = { path: string[]; type: ChangeType };

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(a: any, b: any, path: string[] = []): Generator<Change> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

class ChangelogGenerator {
  addedOrRemovedPallets: Record<string, 'added' | 'removed'> = {};

  changedEvents: Record<
    `${string}.${string}`,
    { field: string; type: 'added' | 'changed' | 'removed' }[]
  > = {};
  addedOrRemovedEvents: Record<`${string}.${string}`, 'added' | 'removed'> = {};

  trackedFieldChanges = new Set<string>();

  add({ path, type }: Change) {
    if (path.length === 1) {
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedPallets[path[0]] = type;
    } else if (path.length === 2) {
      const name = `${path[0]}.${path[1]}` as const;
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedEvents[name] = type;
    } else {
      const [pallet, event, , field] = path;
      const id = `${pallet}.${event}.${field}`;
      if (this.trackedFieldChanges.has(id)) return;
      this.trackedFieldChanges.add(id);
      (this.changedEvents[`${pallet}.${event}`] ??= []).push({ field, type });
    }
  }

  toString() {
    const lines: string[] = [];

    const pallets = Object.keys(this.addedOrRemovedPallets).sort();

    if (pallets.length) {
      lines.push('New or removed pallets:');
      for (const pallet of pallets) {
        lines.push(`  ${pallet}: ${this.addedOrRemovedPallets[pallet]}`);
      }
      lines.push('');
    }

    const events = Object.keys(this.addedOrRemovedEvents)
      .concat(Object.keys(this.changedEvents))
      .sort() as `${string}.${string}`[];

    let currentPallet = '';
    for (const event of events) {
      const [pallet, name] = event.split('.');
      if (pallet !== currentPallet) {
        if (currentPallet) lines.push('');
        currentPallet = pallet;
        lines.push(`${pallet}:`);
      }
      if (event in this.addedOrRemovedEvents) {
        lines.push(`  - ${name}: ${this.addedOrRemovedEvents[event]}`);
      } else {
        lines.push(`  - ${name}:`);
        const fields = this.changedEvents[event].map(
          ({ field, type }) => `    - ${field}: ${type}`,
        );
        lines.push(...fields);
      }
    }

    lines.push('');
    return lines.join('\n');
  }
}

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();
  const changelog = new ChangelogGenerator();

  for (const change of diff(a, b)) {
    changelog.add(change);

    const [pallet, event, field] = change.path;

    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return { changedOrAddedEvents, changelog: changelog.toString() };
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });

const cacheSchema = z.record(
  z.object({
    hash: z.string(),
    network: z.enum(['mainnet', 'perseverance', 'sisyphos', 'backspin', 'localnet']),
  }),
);
export type Network = z.output<typeof cacheSchema>[string]['network'];

export class SpecVersionCache {
  private contents?: z.output<typeof cacheSchema>;

  private readonly queue = new Queue();

  constructor(private readonly path: string) {}

  async read(): Promise<z.output<typeof cacheSchema>> {
    const contents = JSON.parse(
      await fs.readFile(this.path, 'utf8').catch(async () => {
        await fs.mkdir(path.dirname(this.path), { recursive: true });
        return '{}';
      }),
    ) as unknown;

    this.contents ??= cacheSchema.parse(contents);

    return this.contents;
  }

  async write(id: number, hash: string, network: Network): Promise<void> {
    return this.queue.enqueue(async () => {
      const data = await this.read();
      data[id] = { hash, network };
      await fs.writeFile(this.path, JSON.stringify(data, null, 2));
    });
  }

  async getVersion(hash: string, network: Network) {
    const data = await this.read();
    const [version] =
      Object.entries(data).find(([, d]) => d.hash === hash && d.network === network) ?? [];
    return version !== undefined ? Number(version) : undefined;
  }
}

export const specVersionCache = new SpecVersionCache(
  path.join(import.meta.dirname, '..', '..', 'metadata', 'specVersion.json'),
);

export const networkToRpcUrl: Record<Network, string> = {
  mainnet: 'https://rpc.mainnet.chainflip.io',
  perseverance: 'https://archive.perseverance.chainflip.io',
  sisyphos: 'https://archive.sisyphos.chainflip.io',
  backspin: 'https://backspin-rpc.staging',
  localnet: 'http://127.0.0.1:9944',
};
