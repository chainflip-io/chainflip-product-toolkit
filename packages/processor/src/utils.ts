import * as path from 'path';
import * as prettier from 'prettier';
import * as fs from 'fs/promises';
import { z } from 'zod';
import { ParsedMetadata } from './Parser';
import { Queue } from '@chainflip/utils/async';

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(
  a: any,
  b: any,
  path: string[] = [],
): Generator<{ path: string[]; type: 'added' | 'removed' | 'changed' }> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();

  for (const change of diff(a, b)) {
    const [pallet, event, field] = change.path;
    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return changedOrAddedEvents;
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });

const cacheSchema = z.record(
  z.object({
    hash: z.string(),
    network: z.enum(['mainnet', 'perseverance', 'sisyphos', 'backspin']),
  }),
);
export type Network = z.output<typeof cacheSchema>[string]['network'];

export class SpecVersionCache {
  private contents?: z.output<typeof cacheSchema>;

  private readonly queue = new Queue();

  constructor(private readonly path: string) {}

  async read(): Promise<z.output<typeof cacheSchema>> {
    const contents = JSON.parse(
      await fs.readFile(this.path, 'utf8').catch(async () => {
        await fs.mkdir(path.dirname(this.path), { recursive: true });
        return '{}';
      }),
    ) as unknown;

    this.contents ??= cacheSchema.parse(contents);

    return this.contents;
  }

  async write(id: number, hash: string, network: Network): Promise<void> {
    return this.queue.enqueue(async () => {
      const data = await this.read();
      data[id] = { hash, network };
      await fs.writeFile(this.path, JSON.stringify(data, null, 2));
    });
  }

  async getVersion(hash: string, network: Network) {
    const data = await this.read();
    const [version] =
      Object.entries(data).find(([, d]) => d.hash === hash && d.network === network) ?? [];
    return version !== undefined ? Number(version) : undefined;
  }
}

export const specVersionCache = new SpecVersionCache(
  path.join(import.meta.dirname, '..', 'metadata', 'specVersion.json'),
);

export const networkToRpcUrl: Record<Network, string> = {
  mainnet: 'https://mainnet-archive.chainflip.io',
  perseverance: 'https://archive.perseverance.chainflip.io',
  sisyphos: 'https://archive.sisyphos.chainflip.io',
  backspin: 'https://backspin-rpc.staging',
};
