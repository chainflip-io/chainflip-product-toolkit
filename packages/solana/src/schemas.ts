import { assert } from '@chainflip/utils/assertion';
import * as base58 from '@chainflip/utils/base58';
import { assetContractId, chainContractId } from '@chainflip/utils/chainflip';
import { POLKADOT_SS58_PREFIX } from '@chainflip/utils/consts';
import * as ss58 from '@chainflip/utils/ss58';
import BN from 'bn.js';
import { z } from 'zod';
import { tryDecodeCfParams } from './scale';

const entries = Object.entries as <T>(o: T) => [keyof T, T[keyof T]][];

const bnSchema = z
  .any()
  .refine((arg) => BN.isBN(arg))
  .transform((bn) => bn.toString());

const swapParams = z
  .object({
    amount: bnSchema,
    dst_chain: z.number(),
    dst_address: z.instanceof(Buffer),
    dst_token: z.number(),
    ccm_parameters: z
      .object({
        message: z.instanceof(Buffer),
        gas_amount: bnSchema,
      })
      .nullable(),
    cf_parameters: z.instanceof(Buffer).transform((buf) => new Uint8Array(buf)),
  })
  .transform((data, ctx) => {
    const asset = entries(assetContractId).find(([, id]) => id === data.dst_token)?.[0];

    let success = true;

    if (!asset) {
      ctx.addIssue({
        message: 'Failed to get destination asset',
        code: 'custom',
        path: ['dst_token'],
      });
      success = false;
    }

    const chain = entries(chainContractId).find(([, id]) => id === data.dst_chain)?.[0];

    if (!chain) {
      ctx.addIssue({
        message: 'Failed to get destination chain',
        code: 'custom',
        path: ['dst_chain'],
      });
      success = false;
    }

    let destinationAddress: string;

    switch (chain) {
      case 'Arbitrum':
      case 'Ethereum':
        destinationAddress = `0x${data.dst_address.toString('hex')}`;
        break;
      case 'Bitcoin':
        destinationAddress = data.dst_address.toString('utf8');
        break;
      case 'Polkadot':
        destinationAddress = ss58.encode({
          data: data.dst_address,
          ss58Format: POLKADOT_SS58_PREFIX,
        });
        break;
      case 'Solana':
        destinationAddress = base58.encode(data.dst_address);
        break;
      default: {
        // already collecting errors, add type assertion to ensure we handle all chains
        const _: undefined = chain;
        assert(!success, 'expected to be in error state');
      }
    }

    const parsedParams = tryDecodeCfParams(data.cf_parameters);

    if (!parsedParams.ok) {
      parsedParams.reason.forEach((e) =>
        ctx.addIssue({
          message: e.message,
          code: 'custom',
          path: ['cf_parameters'],
        }),
      );
      success = false;
    }

    if (!success) return z.NEVER;

    return {
      depositAmount: data.amount,
      destinationAddress: destinationAddress!,
      destinationChain: chain,
      destinationAsset: asset,
      ccmParams: data.ccm_parameters && {
        message: data.ccm_parameters.message.toString('hex'),
        gasAmount: data.ccm_parameters.gas_amount,
      },
      cfParams: parsedParams.value!,
    };
  });

const swapNativeSchema = z.object({
  name: z.literal('x_swap_native'),
  data: z.object({ swap_native_params: swapParams }).transform((d) => d.swap_native_params),
});

const swapTokenSchema = z.object({
  name: z.literal('x_swap_token'),
  data: z.object({ swap_token_params: swapParams }).transform((d) => d.swap_token_params),
});

export const swapSchema = z.discriminatedUnion('name', [swapNativeSchema, swapTokenSchema]);
