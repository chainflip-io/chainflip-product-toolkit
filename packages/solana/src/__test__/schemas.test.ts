import * as base58 from '@chainflip/utils/base58';
import { assetContractId, chainContractId } from '@chainflip/utils/chainflip';
import * as ss58 from '@chainflip/utils/ss58';
import { BN } from '@coral-xyz/anchor';
import { describe, expect, it } from 'vitest';
import { type z } from 'zod';
import { swapSchema } from '../schemas';

describe('swapSchema', () => {
  it.each([
    ['Bitcoin', assetContractId.Btc, 'bc1qqt3juqef9azhd0zeuamu9c30pg5xdllvmks2ja'],
    [
      'Polkadot',
      assetContractId.Dot,
      ss58.decode('1BzDB5n2rfSJwvuCW9deKY9XnUyys8Gy44SoX8tRNDCFBhx').data,
    ],
    [
      'Solana',
      assetContractId.SolUsdc,
      base58.decode('3yKDHJgzS2GbZB9qruoadRYtq8597HZifnRju7fHpdRC'),
    ],
  ] as const)('handles %s destinations', (chain, asset, address) => {
    const result = swapSchema.parse({
      name: 'x_swap_native',
      data: {
        swap_native_params: {
          // @ts-expect-error -- it's fine
          dst_address: Buffer.from(address),
          dst_chain: chainContractId[chain],
          dst_token: asset,
          amount: new BN(100000000),
          ccm_parameters: {
            message: Buffer.from('deadbeef', 'hex'),
            gas_amount: new BN(10),
          },
          cf_parameters: Buffer.from(
            '00640000002c247dc239834feac0ecdbc5070a7e0b5d0f462d09e9eba871ad1eb9770448234702e76ed292999b076182db5942de83acb5280100000000000000000000000000009e8d88ae895c9b37b2dead9757a3452f7c2299704d91ddfa444d87723f94fe0c000000',
            'hex',
          ),
        },
      },
    } as z.input<typeof swapSchema>);
    expect(result.data).toMatchSnapshot();
  });

  it('collects all the errors', () => {
    expect(() =>
      swapSchema.parse({
        name: 'x_swap_native',
        data: {
          swap_native_params: {
            amount: new BN(100000000),
            dst_chain: 0,
            dst_token: 0,
            dst_address: Buffer.from(''),
            cf_parameters: Buffer.from(''),
            ccm_parameters: null,
          },
        },
      } as z.input<typeof swapSchema>),
    ).toThrowErrorMatchingInlineSnapshot(`
      [ZodError: [
        {
          "message": "Failed to get destination asset",
          "code": "custom",
          "path": [
            "data",
            "swap_native_params",
            "dst_token"
          ]
        },
        {
          "message": "Failed to get destination chain",
          "code": "custom",
          "path": [
            "data",
            "swap_native_params",
            "dst_chain"
          ]
        },
        {
          "message": "Offset is outside the bounds of the DataView",
          "code": "custom",
          "path": [
            "data",
            "swap_native_params",
            "cf_parameters"
          ]
        },
        {
          "message": "Offset is outside the bounds of the DataView",
          "code": "custom",
          "path": [
            "data",
            "swap_native_params",
            "cf_parameters"
          ]
        }
      ]]
    `);
  });
});
