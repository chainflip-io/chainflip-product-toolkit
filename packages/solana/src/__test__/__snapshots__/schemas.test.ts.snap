// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swapSchema > handles Bitcoin destinations 1`] = `
{
  "ccmParams": {
    "gasAmount": "10",
    "message": "deadbeef",
  },
  "cfParams": {
    "affiliateFees": [],
    "boostFee": 0,
    "brokerFees": {
      "account": "cFMTNSQQVfBo2HqtekvhLPfZY764kuJDVFG1EvnnDGYxc3LRW",
      "commissionBps": 0,
    },
    "ccmAdditionalData": null,
    "dcaParams": null,
    "refundParams": {
      "minPriceX128": "6616846606368726564647178815965546002365481543",
      "refundAddress": "3yKDHJgzS2GbZB9qruoadRYtq8597HZifnRju7fHpdRC",
      "retryDurationBlocks": 100,
    },
  },
  "depositAmount": "100000000",
  "destinationAddress": "bc1qqt3juqef9azhd0zeuamu9c30pg5xdllvmks2ja",
  "destinationAsset": "Btc",
  "destinationChain": "Bitcoin",
}
`;

exports[`swapSchema > handles Polkadot destinations 1`] = `
{
  "ccmParams": {
    "gasAmount": "10",
    "message": "deadbeef",
  },
  "cfParams": {
    "affiliateFees": [],
    "boostFee": 0,
    "brokerFees": {
      "account": "cFMTNSQQVfBo2HqtekvhLPfZY764kuJDVFG1EvnnDGYxc3LRW",
      "commissionBps": 0,
    },
    "ccmAdditionalData": null,
    "dcaParams": null,
    "refundParams": {
      "minPriceX128": "6616846606368726564647178815965546002365481543",
      "refundAddress": "3yKDHJgzS2GbZB9qruoadRYtq8597HZifnRju7fHpdRC",
      "retryDurationBlocks": 100,
    },
  },
  "depositAmount": "100000000",
  "destinationAddress": "1BzDB5n2rfSJwvuCW9deKY9XnUyys8Gy44SoX8tRNDCFBhx",
  "destinationAsset": "Dot",
  "destinationChain": "Polkadot",
}
`;

exports[`swapSchema > handles Solana destinations 1`] = `
{
  "ccmParams": {
    "gasAmount": "10",
    "message": "deadbeef",
  },
  "cfParams": {
    "affiliateFees": [],
    "boostFee": 0,
    "brokerFees": {
      "account": "cFMTNSQQVfBo2HqtekvhLPfZY764kuJDVFG1EvnnDGYxc3LRW",
      "commissionBps": 0,
    },
    "ccmAdditionalData": null,
    "dcaParams": null,
    "refundParams": {
      "minPriceX128": "6616846606368726564647178815965546002365481543",
      "refundAddress": "3yKDHJgzS2GbZB9qruoadRYtq8597HZifnRju7fHpdRC",
      "retryDurationBlocks": 100,
    },
  },
  "depositAmount": "100000000",
  "destinationAddress": "3yKDHJgzS2GbZB9qruoadRYtq8597HZifnRju7fHpdRC",
  "destinationAsset": "SolUsdc",
  "destinationChain": "Solana",
}
`;
